WT-12412 Model functions set_stable_timestamp and set_oldest_timestamp should match WT behavior

pmacko86 opened the pull request.

korteland commented on test/model/test/model_basic/main.cpp:
```diff
@@ -1223,9 +1222,8 @@ test_model_oldest_wt(void)
-    /* Test moving the oldest timestamp backwards - this should fail silently. */
-    /* wt_model_set_oldest_timestamp_both(10); */
+    /* Test moving the oldest timestamp backwards - this should fail. */
+    wt_model_set_oldest_timestamp_both(10);
```
Is this call missing a `testutil_check(... == EINVAL)` wrapper?

pmacko86 replied:
The definition of `wt_model_set_oldest_timestamp_both` includes an assertion that both WT and the model return the same return value.

korteland replied:
Got it - thanks.

korteland commented on test/model/test/model_checkpoint/main.cpp:
```diff
@@ -284,10 +284,8 @@ test_checkpoint_wt(void)
-     */
-    wt_model_set_stable_timestamp_both(60);
+    /* Test moving the stable timestamp backwards - this should fail. */
+    wt_model_set_stable_timestamp_both(50);
```
Same deal - `testutil_check`?

pmacko86 replied:
The same for `wt_model_set_stable_timestamp_both`.

korteland reviewed (approved):
LGTM :-)

etienneptl commented on test/model/test/model_checkpoint/main.cpp:
```diff
@@ -157,22 +157,22 @@ test_checkpoint(void)
     testutil_assert(table->get(ckpt3, key2) == value2); /* The old value. */
     testutil_assert(table->get(ckpt3, key3) == value3);
 
     /* Test moving the stable timestamp backwards - this should fail silently. */
```
It no longer fails "silently":
```suggestion
    /* Test moving the stable timestamp backwards - this should fail. */
```

pmacko86 replied:
Nice catch. Fixed in [2ca293c](https://github.com/wiredtiger/wiredtiger/pull/10545/commits/2ca293cc206a4fcd452d44580c27e201e2eeab10). Thanks!

etienneptl reviewed (approved):
LGTM, left one comment.
Do you have a scenario where the oldest is set before the stable is? This is the only scenario where we can have an oldest ts "greater" than the stable ts because it is unset.

pmacko86 commented:
> Do you have a scenario where the oldest is set before the stable is? This is the only scenario where we can have an oldest ts "greater" than the stable ts because it is unset.

Yes, in the test `model_basic`, in `test_model_oldest()` and `test_model_oldest_wt()`:

```
    /* Now try it with both the oldest and stable timestamps. */
    wt_model_set_oldest_timestamp_both(50);
    wt_model_set_stable_timestamp_both(55);
    database.restart();
    testutil_check(session->close(session, nullptr));
    testutil_check(conn->close(conn, nullptr));
    testutil_wiredtiger_open(opts, test_home.c_str(), ENV_CONFIG, nullptr, &conn, false, false);
    testutil_check(conn->open_session(conn, nullptr, nullptr, &session));
    testutil_assert(database.oldest_timestamp() == wt_get_oldest_timestamp(conn));
```

pmacko86 commented:
Oh, I see, never mind. There are no such relative checks for the oldest and the stable timestamp, so I just added them in; that's a good suggestion. I just added them within this ticket as it's close enough to its scope; see [16aeda8](https://github.com/wiredtiger/wiredtiger/pull/10545/commits/16aeda82ee7d355bdf0b9f9a57bad3b8064576ff).

@etienneptl, @korteland, given that this is a non-trivial (albeit small) change, could you please review this last commit before I merge?

korteland reviewed (approved):
Still looks good.

etienneptl reviewed (approved):
LGTM! As always, thank you for taking the time to address all the comments/questions!
